tensorflow_probability/python/experimental/mcmc/streaming_reductions.py

```diff
@@ -72,7 +72,10 @@ def step_kernel(
       `return_final_kernel_results` is `True`.
   """
   is_seeded = seed is not None
-  seed = samplers.sanitize_seed(seed, salt='mcmc.sample_chain')
+  if tf.executing_eagerly():
+    tf.random.set_seed(seed)
```

Comment:
Definitely don't do this.

I'm afraid we didn't actually explain what our prng seeding semantics actually are, which is why you got confused.  There are actually two seed types:
- A Python integer, which triggers "Classic TF" seed semantics, and
- A Tensor of 2 int32, which triggers "stateless" seed semantics.

The stateless ones are simpler: Any random function from TFP, and from tf.random.stateless_*, when given a seed of that type, is just a deterministic function of the seed.

The Classic TF ones are complicated, and use mysteriously changing hidden state to try to approximate the "mutating PNG" style that one gets from, say, numpy.  Unfortunately, the resulting semantics are very complicated in graph mode, don't correctly support parallel iterations in while loops, are different in eager mode, and (IIRC) don't work with XLA at all (because XLA doesn't support the required hidden changing state).

TFP supports the Classic TF semantics for backward compatibility, but all of our code (at least that pertains to MCMC) is internally coded in stateless style, so we convert incoming seeds to the type that triggers stateless behavior on input to every public function.

That's what `sanitize_seed` does: if the incoming seed is of stateless type (i.e., a Tensor of two int32), it just passes it along (possibly with salting, though that's irrelevant).  But if the incoming seed is of Classic TF type, i.e. the user asked for Classic TF sampling semantics, then `sanitize_seed` invokes `tf.random.uniform` (which has Classic TF seeding semantics) to generate a stateless seed.  Thus, `sanitize_seed(7)` is not a deterministic function of 7, per the complicated Classic TF rules; but if you only call `sanitize_seed(7)` once in a given test, the returned object is a seed that will trigger the simple, deterministic prng semantics we all know and love.

The last piece of this is that `test_util.test_seed()` returns a Classic TF seed (i.e., a Python integer), because it's older than the migration to stateless seeds.  Maybe it's getting to be time to fix that (or at least add another test utility).

Bottom line: a modern reproducibility test should look like this:
```
def test_repro_foo_stateless(self):
  seed = samplers.sanitize_seed(test_util.test_seed())
  ans1_ = something_random(seed=seed)
  ans2_ = something_random(seed=seed)
  ans1, ans2 = self.evaluate([ans1, ans2])
  self.assertEqual(ans1, ans2)
```

There is also a pattern for reproducibility tests for TF Classic seeding semantics, but I don't think it's important for this PR (or much in general, as long as we maintain the discipline of sanitizing seeds).

```suggestion
```

Reply:
Thanks so much for the in-depth explanation! Everything makes a lot more sense now haha